lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,308 @@
-      foundOperators.push_back(cast<FuncDecl>(VD));
-      continue;
-    }
+static std::string escape(const DeclBaseName name) {
+  return llvm::yaml::escape(name.userFacingName());
+}
 
-    auto nominal = dyn_cast<NominalTypeDecl>(D);
-    if (!nominal)
-      continue;
-    foundNominals[nominal] |= true;
-    findNominalsAndOperators(foundNominals, foundOperators,
-                             nominal->getMembers());
-  }
+static std::string mangleTypeAsContext(const NominalTypeDecl *type) {
+  Mangle::ASTMangler Mangler;
+  return Mangler.mangleTypeAsContextUSR(type);
 }
 
-static bool declIsPrivate(const Decl *member) {
-  auto *VD = dyn_cast<ValueDecl>(member);
-  if (!VD) {
-    switch (member->getKind()) {
-    case DeclKind::Import:
-    case DeclKind::PatternBinding:
-    case DeclKind::EnumCase:
-    case DeclKind::TopLevelCode:
-    case DeclKind::IfConfig:
-    case DeclKind::PoundDiagnostic:
-      return true;
+template <typename Decl_T>
+static bool hasSomePrivateAccessLevel(const Decl_T *decl) {
+  return decl->hasAccess() &&
+         decl->getFormalAccess() <= AccessLevel::FilePrivate;
+}
 
-    case DeclKind::Extension:
-    case DeclKind::InfixOperator:
-    case DeclKind::PrefixOperator:
-    case DeclKind::PostfixOperator:
-      return false;
+/// If this declaration's access level confines it to a file, no other file can
+/// depend upon it.
+template <typename Decl_T>
+static bool isVisibleOutsideItsFile(const Decl_T *decl) {
+  return !hasSomePrivateAccessLevel(decl);
+}
 
-    default:
-      llvm_unreachable("everything else is a ValueDecl");
+class ReferenceDependencyFileEmitter {
+  SourceFile *SF;
+  const DependencyTracker &depTracker;
+  std::unique_ptr<llvm::raw_fd_ostream> out;
+
+  ReferenceDependencyFileEmitter(SourceFile *SF,
+                                 const DependencyTracker &depTracker,
+                                 std::unique_ptr<llvm::raw_fd_ostream> out)
+      : SF(SF), depTracker(depTracker), out(std::move(out)) {
+    assert(SF && "Cannot emit reference dependencies without a SourceFile");
+  }
+
+public:
+  static Optional<ReferenceDependencyFileEmitter>
+  create(DiagnosticEngine &diags, SourceFile *SF,
+         const DependencyTracker &depTracker, StringRef outputPath) {
+    // Before writing to the dependencies file path, preserve any previous file
+    // that may have been there. No error handling -- this is just a nicety, it
+    // doesn't matter if it fails.
+    llvm::sys::fs::rename(outputPath, outputPath + "~");
+
+    std::error_code EC;
+    auto out = llvm::make_unique<llvm::raw_fd_ostream>(outputPath, EC,
+                                                       llvm::sys::fs::F_None);
+
+    if (out->has_error() || EC) {
+      diags.diagnose(SourceLoc(), diag::error_opening_output, outputPath,
+                     EC.message());
+      out->clear_error();
+      return None;
     }
+
+    return ReferenceDependencyFileEmitter(SF, depTracker, std::move(out));
```

Comment:
I'm not sure you really need a return value at all. You could just rename this to `emit` and have it return a success/failure boolean (or the std::error_code, maybe).

Reply:
Yes! Nice!!